cpp/src/io/utilities/time_utils.cuh

```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+namespace cudf {
+namespace io {
+
+/**
+ * @brief Function that translates cuDF time unit to clock frequency
+ */
+constexpr int32_t to_clockrate(type_id timestamp_type_id)
+{
+  switch (timestamp_type_id) {
+    case type_id::DURATION_SECONDS:
+    case type_id::TIMESTAMP_SECONDS: return timestamp_s::period::den;
+    case type_id::DURATION_MILLISECONDS:
+    case type_id::TIMESTAMP_MILLISECONDS: return timestamp_ms::period::den;
+    case type_id::DURATION_MICROSECONDS:
+    case type_id::TIMESTAMP_MICROSECONDS: return timestamp_us::period::den;
+    case type_id::DURATION_NANOSECONDS:
+    case type_id::TIMESTAMP_NANOSECONDS: return timestamp_ns::period::den;
+    default: return 0;
+  }
```

Comment:
This should really be type dispatched.
```suggestion
struct get_period{
   template <typename T>
   auto operator()(){
      if constexpr(is_chrono<T>()){
         return T::period::den;
      }
      CUDF_FAIL("Invalid, non chrono type");
   }
};


constexpr int32_t to_clockrate(type_id timestamp_type_id)
{
   return type_dispatcher(timestamp_type_id, get_period{});
}
```

Reply:
```cpp
#include <cudf/types.hpp>
#include <cudf/utilities/type_dispatcher.hpp>
#include <cudf/wrappers/timestamps.hpp>

struct get_period {
  template <typename T>
  constexpr auto operator()() const
  {
    if constexpr (cuda::std::chrono::is_chrono<T>()) { return T::period::den; }
    CUDF_FAIL("Invalid, non chrono type");
  }
};

/**
 * @brief Function that translates cuDF time unit to clock frequency
 */
constexpr int32_t to_clockrate(type_id timestamp_type_id)
{
  return cudf::type_dispatcher(timestamp_type_id, get_period{});
}
```
Here's the error I got:
```
/home/yunsongw/dev/rapids/cudf/cpp/src/io/utilities/time_utils.cuh:64: error: no instance of function template "cudf::type_dispatcher" matches the argument list
            argument types are: (cudf::type_id, cudf::io::get_period)
```
Any idea why the above code is not working?

Reply:
Just realized `orc::TIMESTAMP` can be `type_id::EMPTY` also and `type_dispatcher` does not handle `type_id::EMPTY`.